internal/healthchecks/builtin.go

```diff
@@ -0,0 +1,200 @@
+				continue
+			}
+		case *appsv1.DaemonSet:
+			if err := client.Get(ctx, objectKey, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.NumberAvailable != obj.Status.DesiredNumberScheduled {
+				gvkErrors = appendResourceError(gvkErrors, obj, "DaemonSet is not ready")
+			}
+		case *appsv1.ReplicaSet:
+			if err := client.Get(ctx, objectKey, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.AvailableReplicas != obj.Status.Replicas {
+				gvkErrors = appendResourceError(gvkErrors, obj, "ReplicaSet is not ready")
+			}
+		case *corev1.Pod:
+			if err := client.Get(ctx, objectKey, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+			if obj.Status.Phase != corev1.PodRunning && obj.Status.Phase != corev1.PodSucceeded {
+				gvkErrors = appendResourceError(gvkErrors, obj, "Pod is not Running or Succeeded")
+			}
+		case *apiregistrationv1.APIService:
+			if err := client.Get(ctx, objectKey, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+
+			conditionExists := false
+			for _, condition := range obj.Status.Conditions {
+				if condition.Type == apiregistrationv1.Available {
+					if condition.Status != "True" {
+						gvkErrors = appendResourceError(gvkErrors, obj, condition.Message)
+					}
+					conditionExists = true
+					break
+				}
+			}
+			if conditionExists {
+				continue
+			}
+			// If we are here we didn't find the "Available" condition, so we assume the APIService is non healthy.
+			gvkErrors = appendResourceError(gvkErrors, obj, "Available condition not found")
+		case *apiextensionsv1.CustomResourceDefinition:
+			if err := client.Get(ctx, objectKey, obj); err != nil {
+				gvkErrors = appendResourceError(gvkErrors, obj, err.Error())
+				continue
+			}
+
+			conditionExists := false
+			for _, condition := range obj.Status.Conditions {
+				if condition.Type == apiextensionsv1.Established {
+					if condition.Status != "True" {
+						gvkErrors = appendResourceError(gvkErrors, obj, condition.Message)
+					}
+					conditionExists = true
+					break
+				}
+			}
+			if conditionExists {
+				continue
+			}
+			gvkErrors = appendResourceError(gvkErrors, obj, "Established condition not found")
+		default:
+			// If we don't know how to check the health of the object, we assume it's healthy.
+			continue
+		}
+	}
+
+	return errors.Join(gvkErrors...)
+}
+
+// toErrKey returns a string that identifies a resource based on its GVK and namespace/name. This key is used
+// to identify the resource in the error message.
+func toErrKey(resource client.Object) string {
+	group := resource.GetObjectKind().GroupVersionKind().Group
+	namespace := resource.GetNamespace()
```

Comment:
In theory, since `resource` is an interface, the underlying implementation is unknown by the caller, so it might make sense to use the `GetNamespace` method just once and reuse the returned string everywhere.

In practice, it probably doesn't make much difference.

Reply:
simplified, done